pallets/manta-pay/src/lib.rs

```diff
@@ -0,0 +1,784 @@
+			}
+		}
+	}
+}
+
+impl<T> TransferLedger<config::Config> for Ledger<T>
+where
+	T: Config,
+{
+	type AccountId = T::AccountId;
+	type UpdateError = FungibleLedgerConsequence<asset::AssetValueType>;
+	type Event = PreprocessedEvent<T>;
+	type ValidSourceAccount = WrapPair<Self::AccountId, asset::AssetValue>;
+	type ValidSinkAccount = WrapPair<Self::AccountId, asset::AssetValue>;
+	type ValidProof = Wrap<()>;
+	type SuperPostingKey = ();
+
+	#[inline]
+	fn check_source_accounts<I>(
+		&self,
+		asset_id: manta_accounting::asset::AssetId,
+		sources: I,
+	) -> Result<Vec<Self::ValidSourceAccount>, InvalidSourceAccount<Self::AccountId>>
+	where
+		I: Iterator<Item = (Self::AccountId, asset::AssetValue)>,
+	{
+		sources
+			.map(move |(account_id, withdraw)| {
+				T::FungibleLedger::can_withdraw(asset_id.0, &account_id, withdraw.0)
+					.and_then(|_| Ok(WrapPair(account_id.clone(), withdraw)))
+					.map_err(|_| InvalidSourceAccount {
+						account_id,
+						asset_id,
+						withdraw,
+					})
+			})
+			.collect()
+	}
+
+	#[inline]
+	fn check_sink_accounts<I>(
+		&self,
+		asset_id: manta_accounting::asset::AssetId,
+		sinks: I,
+	) -> Result<Vec<Self::ValidSinkAccount>, InvalidSinkAccount<Self::AccountId>>
+	where
+		I: Iterator<Item = (Self::AccountId, asset::AssetValue)>,
+	{
+		// NOTE: Existence of accounts is type-checked so we don't need to do anything here, just
+		//		 pass the data forward.
+		sinks
+			.map(move |(account_id, deposit)| {
+				T::FungibleLedger::can_deposit(asset_id.0, &account_id, deposit.0)
+					.and_then(|_| Ok(WrapPair(account_id.clone(), deposit)))
+					.map_err(|_| InvalidSinkAccount {
+						account_id,
+						asset_id,
+						deposit,
+					})
+			})
+			.collect()
+	}
+
+	#[inline]
+	fn is_valid(
+		&self,
+		asset_id: Option<asset::AssetId>,
+		sources: &[SourcePostingKey<config::Config, Self>],
+		senders: &[SenderPostingKey<config::Config, Self>],
+		receivers: &[ReceiverPostingKey<config::Config, Self>],
+		sinks: &[SinkPostingKey<config::Config, Self>],
+		proof: Proof<config::Config>,
+	) -> Option<(Self::ValidProof, Self::Event)> {
+		let (mut verifying_context, event) = match TransferShape::select(
+			asset_id.is_some(),
+			sources.len(),
+			senders.len(),
+			receivers.len(),
+			sinks.len(),
+		)? {
+			TransferShape::Mint => (
+				manta_sdk::pay::testnet::verifying::Mint::get().expect("Checksum did not match."),
+				PreprocessedEvent::<T>::Mint {
+					asset: Asset::new(asset_id.unwrap().0, (sources[0].1).0),
+					source: sources[0].0.clone(),
+				},
+			),
+			TransferShape::PrivateTransfer => (
+				manta_sdk::pay::testnet::verifying::PrivateTransfer::get()
+					.expect("Checksum did not match."),
+				PreprocessedEvent::<T>::PrivateTransfer,
+			),
+			TransferShape::Reclaim => (
+				manta_sdk::pay::testnet::verifying::Reclaim::get()
+					.expect("Checksum did not match."),
+				PreprocessedEvent::<T>::Reclaim {
+					asset: Asset::new(asset_id.unwrap().0, (sinks[0].1).0),
```

Comment:
Remove `unwrap`?

Reply:
unfortunately, I don't think this one can be removed with the current design.